### DISCUSSION TITLE
Typing `createAsyncThunk` with JSDoc?

### DISCUSSION THREAD

**Discussion by Arkellys** [reduxjs/redux-toolkit#4437] [Q&A]
I'm trying to write types with JSDoc for an action created with `createAsyncThunk`, but I can't find a way to get the correct return type for `getState`. I get the expected type `ParamType` for `params` but `getState` always return `unknown`:

```js
/**
 * @import { AsyncThunk } from "@reduxjs/toolkit"
 * @type {AsyncThunk<any, ParamType, { state: AppState }>}
 */

export const myAction = createAsyncThunk(
    ACTIONS.MY_ACTION,
    async (params, { getState }) => {
        const state = getState() // returns type `unknown`
        // ...
    },
);
```

As far as I know, what I pass seems correct. `AppState` is an object, and I also tried with dummy objects and strings to be sure it was not the problem, but I still get `unknown`. 

I'm aware of the documentation of Redux usage with TS, but I don't use TS, and I can't understand why writing the type like this works for everything but the custom `ThunkApiConfig`.

**Comment by EskiMojo14** [comment]
I think your best bet is annotating the payload creator instead:

```ts
export const myAction = createAsyncThunk(
    ACTIONS.MY_ACTION,
    /** @type {import("@reduxjs/toolkit").AsyncThunkPayloadCreator<any, ParamType, { state: AppState }>}
    async (params, { getState }) => {
        const state = getState() // returns type `AppState`
        // ...
    },
);
```

**Reply by Arkellys** [reply]
Thank you, but somehow if it works with an inline import, I still get an error when importing via `@import`. The return type of `AsyncThunkPayloadCreator` is `MaybePromise`:

```js
type MaybePromise<T> = T | Promise<T> | (T extends any ? Promise<T> : never);
```
And this fails on `async` functions:

```js
The return type of an async function or method must be the global Promise<T> type.
```

I don't know why the error do not show with inline imports, but the tag import does show it and from what I read [here](https://stackoverflow.com/questions/57695992/in-typescript-why-cant-an-async-function-return-a-union-of-type-t-promiset), it seems to be in the right.

So it's just replacing an error with another, and I have no way of overriding the types. All of these look like bugs to me. Why using `AsyncThunk` doesn't work in the first place?

**Reply by EskiMojo14** [reply]
Because annotating AsyncThunk like that can't affect the inference of the AsyncThunkConfig. It just hits the first overload with no AsyncThunkConfig specified.

![image](https://github.com/reduxjs/redux-toolkit/assets/18308300/8ad87473-b98c-492d-b731-b4aadd9952b4)

That's not a bug, that's just how the types work.

**Reply by Arkellys** [reply]
Well... I wasted too much hours because of this, so I ended up creating a `postinstall` script to fix the types of `createAsyncThunk`. It's bad to do that, but seriously, typing Redux actions shouldn't be so complicated.

If someone need it:

```js
const fs = require("fs");
const { noop } = require("lodash");
const path = require("path");


const reduxToolkitPath = "node_modules/@reduxjs/toolkit/dist";
const createAsyncThunkTypeFileName = "createAsyncThunk.d.ts";
const createAsyncThunkTypeFilePath = path.join(reduxToolkitPath, createAsyncThunkTypeFileName);

// No logs because Yarn does not print logs from `postinstall` script

fs.readFile(createAsyncThunkTypeFilePath, "utf8", (_error, data) => {
  const badTypeRegex = /<Returned, ThunkArg = void>.+>;/;

  const dataWithFixedTypes = data.replace(badTypeRegex, (match) => {
    const matchWithNewParam = match.replace("void>", "void, ThunkApiConfig extends AsyncThunkConfig>");
    return matchWithNewParam.replace(/CurriedThunkApiConfig/g, "ThunkApiConfig");
  });

  fs.writeFile(createAsyncThunkTypeFilePath, dataWithFixedTypes, "utf8", noop);
});
```

Thank you for your help.

**Reply by EskiMojo14** [reply]
fwiw there are packages like [patch-package](https://www.npmjs.com/package/patch-package) or [yarn patch](https://yarnpkg.com/cli/patch) that probably make that process a little easier

**Reply by Arkellys** [reply]
Thank you, I will take a look!

**Reply by phryneas** [reply]
Uhh... You know that yarn has [`yarn patch`](https://yarnpkg.com/cli/patch) built in?

Generally: please tone the snarkyness down. It's frustrating, and doesn't really encourage us to help you.
 
By choosing JSDoc you have deliberately chosen the harder path of doing things. That's up to you - but it's not our fault. 
Our types are optimized to give TypeScript users a good editor experience. We don't ship them as `.d.jsdoc`, but as `.d.ts`, because they're meant to be used with TypeScript.
 
Also, as far as I understand, @EskiMojo14's suggestion up there works, and you just don't like the syntax. Again, it seems like you are deliberately making things hard for yourself here.

---

Going back to your question: Even if you were using TypeScript, you're just using it wrong. TypeScript will never infer the type of a method argument from typing the return value. It only works the other way round.

The suggested way of doing this in TypeScript would be

```ts
import { createAsyncThunk } from '@reduxjs/toolkit'

// usually, you would do this once in your application
const catWithState = createAsyncThunk.withTypes<{ state: {foo: "bar"} }>()

const thunk = catWithState('foo', (arg: undefined, {getState} ) => {
  const value = getState()
  //      ^? - is { foo: "bar" }
})
```

I honestly have no idea on how to write that with JSDoc, but again, you've deliberately chosen the hard way of doing things.

**Reply by Arkellys** [reply]
@phryneas I'm very sorry if my tone was "snarky", it's not against the lib or people here. I spent too much time trying to solve this problem without taking a break, and too much time figuring out where the problem was. It definitely frustrated me. 😅

I didn't know about `yarn patch`, I've discovered it with the message of EskiMojo14, and I will clearly take a look at it!

As I said, using `AsyncThunkPayloadCreator` doesn't work with `async` function. The inline import may not throw the error, but using `@import` does, and [TS playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAsghiARhACgJwPYFsCWBnCAHgBUA+KAXimKgB8p1t8iy6oAKGiAD2AgDsAJnihx+IKAH4GmXARLkAXFH4QAbhDQBKANwAoPQGMM-PMCh8zAQTwh+hyqNv2OW5fCSpZzQmbQ5+AHNyCnIAbz0oKGNTc0tzKjgAdzgcc1UkmSZ5PwDg9jQIPAwAGw1KckLisoh2ACJ4uq0tSKhC4ABXNH4VCEzGOVqq0vLQtqKR2vjm-QBfAxizCyLgADkMGzsHKld3BGQBn1ygkPDW9q6ejKzBgomaivHqjXrGmb1ZoA) shows the error is correct. Appart from not liking the synthax, the inline import also hides an error. I'm not expert on TS, so maybe I've just misunderstood something?

AFAIK, I cannot use `withTypes` with JSDoc. But all I need is the first overload of `createAsyncThunk` to accept `ThunkApiConfig` as third param. So from this:

```ts
<Returned, ThunkArg = void>(
    typePrefix: string,
    payloadCreator: AsyncThunkPayloadCreator<
      Returned,
      ThunkArg,
      CurriedThunkApiConfig
    >,
    options?: AsyncThunkOptions<ThunkArg, CurriedThunkApiConfig>,
  ): AsyncThunk<Returned, ThunkArg, CurriedThunkApiConfig>
```

To this:

```ts
<Returned, ThunkArg = void, ThunkApiConfig extends AsyncThunkConfig>(
    typePrefix: string,
    payloadCreator: AsyncThunkPayloadCreator<
      Returned,
      ThunkArg,
      ThunkApiConfig
    >,
    options?: AsyncThunkOptions<ThunkArg, ThunkApiConfig>,
  ): AsyncThunk<Returned, ThunkArg, ThunkApiConfig>
```

And now everything works like a charm with:

```js
/**
 * @import { AsyncThunk } from "@reduxjs/toolkit"
 * @type {AsyncThunk<any, ParamType, { state: AppState }>}
 */
```

TS flavored JSDoc is a very nice alternative to get typing without needing a transpiler. I don't think it's the hard way, I've managed to do some pretty complex things with it. So far typing Redux actions is the only thing I've really struggled with, it's indeed not made to support JSDoc. 🥲

**Reply by EskiMojo14** [reply]
pretty nasty, but you can simulate calling withTypes like this:
```ts
import { createAsyncThunk } from "@reduxjs/toolkit";

/** @type {ReturnType<typeof createAsyncThunk.withTypes<{ state: "foo" }>>} */
const createWithTypes = createAsyncThunk
```

**Reply by Arkellys** [reply]
Interesting, I didn't know it was possible to do that! I may give it a try and see what I can do or if the patch is easier. Thank you again. 🙂

**Reply by markerikson** [reply]
@Arkellys : 

> typing Redux actions shouldn't be so complicated

I'll echo Lenz's comments. The RTK types do work well, when used as intended. You've made the entire process much harder on yourself.

There's always room for legitimate improvements, and we do welcome suggestions, but complaining "this is hard and doesn't work right" when you're _not using it the way we intended_ is not helpful.

**Reply by Arkellys** [reply]
But I never said it was not working right, and I made it clear I was trying to type with JSDoc instead of TS and that I was aware there was a doc made for TS. I did said it was looking like bugs to me because I could see `AsyncThunk` was taking a `ThunkApiConfig`, but I couldn't understand why it was ignoring it. That's why I asked in this discussion. I was genuinely just trying to understand if it was a bug and how I could fix it on my side, I never expected for the lib to do it.
